sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,220 @@
+     * @return next work to process from queue.
+     */
+    private SynchronousReceiveWork getNextWorkAndRequest() {
+        // Now see if there is more queued up work
+        SynchronousReceiveWork nextWork = workQueue.poll();
+
+        if (nextWork != null) {
+
+            logger.verbose("[{}] Picking up next receive request.", nextWork.getId());
+
+            // timer to complete the current in case of timeout trigger
+            timeoutOperation = getTimeoutOperation(nextWork.getTimeout());
+
+            requestCredits(nextWork.getNumberOfEvents());
+        }
+        return nextWork;
+    }
     /**
      * Publishes the event to the current {@link SynchronousReceiveWork}. If that work item is complete, will dispose of
      * the subscriber.
-     *
-     * @param value Event to publish.
+     * @param message Event to publish.
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
+        boolean delivered = false;
+
+        // Boundary condition(timeout case): if timeout has happened, get next available work.
+        if (currentWork == null) {
+            // Now see if there is more queued up work
+            currentWork = getNextWorkAndRequest();
+            logger.verbose("No current work, Picked up next receive request.");
+        }
+
+        if (currentWork != null) {
+            currentWork.next(message);
+            delivered = true;
+            remaining.decrementAndGet();
+            // Check if we have delivered all the messages to current work.
+            if (currentWork.isTerminal()) {
+                currentWork.complete();
+                if (timeoutOperation != null && !timeoutOperation.isDisposed()) {
+                    timeoutOperation.dispose();
+                }
+
+                // Now see if there is more queued up work
+                currentWork = getNextWorkAndRequest();
+                logger.verbose("Current work is terminal, Picked up next receive request.");
+            }
+        }
 
-        if (work.isTerminal()) {
-            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", work.getId());
-            dispose();
+        if (currentWork == null) {
+            if (wip.decrementAndGet() != 0) {
+                logger.warning("There is another worker in drainLoop. But there should only be 1 worker.");
+            }
+            // If message is not delivered to downstream, we will buffer it.
+            if (!delivered) {
+                bufferMessages.add(message);
+            }
         }
     }
 
-    @Override
-    protected void hookOnComplete() {
-        logger.info("[{}] Completed. No events to listen to.", work.getId());
-        dispose();
+    /**
+     *
+     * @param requested credits for current {@link SynchronousReceiveWork}.
+     */
+    private void requestCredits(long requested) {
```

Comment:
In general, keep the methods with the same visibility together. I organise visibility of methods in descending visibility (ie. public, package-private, private) because consumers will interact with the public/package-private ones first.

As-is, I need to jump around code to reason about what you are trying to do.